### ISSUE TITLE
Blind users costing model

### ISSUE THREAD

**Issue by grzezlo** [valhalla/valhalla#2989]
Many people with blindness are using navigational mobile apps specially designed for them, providing area or route informations as text read by TTS, or screen-reading software installed in mobile device, instead of map rendered visually on the screen.
Would it be possible to introduce "blind users" costing model in Valhalla, which would make it even more perfect as a routing engine for this group of end-users?
"Blind users" costing model could be basically the pedestrian model, with some extensions:
1. Announce in routing instructions, streets crossed by the user with the hint to cross it and go forward.
2. Generate instruction, when user should change the street side.
3. If possible, announce the type of pedestrian crossing: if has  traffic lights, even better if has tag traffic_signals:sound=yes, or traffic_signals:vibration=yes.
4. Some further improvements and optimizations would be possible in this model, e.g. costing_options to avoid huge streets with no signals if crossing needed, and so on, but not all can be possible to implement without changing the data structures, and even in the basic form, with points 1 and 2, it would be great help.

Possibly related to [#2942: Pedestrian crossing should be a separate TripLeg_Use ](https://github.com/valhalla/valhalla/issues/2942)

**Comment by kevinkreiser** [comment]
@grzezlo sorry for the delay in reply. 

The extensions you list 1-3 are all related to Odin, the part of the code that decides maneuvers and directions to call out. So I think no modifications to the costing model will be needed for those. I think we could take care of those in Odin by adding options for those to the request though, which i think is more what you were getting at.

1. this one is possible, Odin does get crossing street information which it uses in determining maneuvers we could add additional "maneuver notes" with cross streets but it seems to me you want actual maneuvers to happen when we cross a street even if it isn't a "turn". this makes sense to me
2. this is quite tricky. at the moment we dont have a notion of what side of the street you are on except at the very beginning and end of the route (basically the start/end is on the left/right side of the street). I believe we could keep track of this information along the path but it requires us to know what side you started on with certainty, which means that in some cases we wont be able to do this (if the start location is really close to the road center line).
3. we should be able to tell the user when there is a traffic light but at the moment we do not parse the sound or vibration tags. We could add those as auxiliary data to edgeinfo similar to how we did for bridge/tunnel names or how we'll do for pronounciations.
4. these changes do seem like costing_options for pedestrian to me. looks like we want to penalize transitions over high class roads with no traffic signal. the problem is we dont get the cross street information (specifically the road class) when we are doing the graph search. we could add this but it will be a performance problem most likely. maybe we could store more information on the node like highest and lowest class edge connected to this node, this would allow us to potentially infer this data very quickly. that would require an additional 6 bits on the baldr::NodeInfo. maybe we have something else that we could use here too.

@dgearhart @dnesbitt61 any comments?

**Comment by grzezlo** [comment]
Thanks @kevinkreiser for your reply and interest in proposed idea.
Generally speaking, everything what would be doable without decreasing the quality of software for other users, would be a great help for blind people, for whom the navigational informations in space are more significant than for others, who easily can look around and corelate navigational informations with seen reality.
Valhalla even currently is great tool, and with proposed improvements could became even greater in described use case.
The different levels of data quality and limited technical possibilities of GPS systems and software decreasing the reliability of generated route are obvious, so every blind man such route has to treat with limited trust, similarly to all other informations got from navigational software.
Difficulties related to 2. are understandable, but maybe some assumptions would be enough: if given road is not a footway or track, and user got instruction to turn left in it, we can assume he is on the left side of the road.
Of course, it can lead to wrong results, so should be further analyzed or tested, but maybe request parameter "assume_street_side", false by default, would solve the problem.
If start location is at the center of line, or very near by, the algorithm could not make assumptions on this particular way, and could start to do it after first turn.
3. Great news, that the traffic lights could be announced, even without the sound or vibration it would be helpful.
4. I understand the complications, I supposed before that it could be problematic.

**Comment by grzezlo** [comment]
Thanks again @kevinkreiser for your reply.
Finally I started to hack the code on my own, and have some interesting first results.
One question: how to get names of streets the user intersects?
The maneuver.cross_street_names() seems to be here for it, but it doesn't seem to be set up anywhere.
Also the tripleg_intersecting_edge class has no name attribute.
Am I missing something?

**Comment by mandeepsandhu** [comment]
CC: @dgearhart

**Comment by dgearhart** [comment]
> One question: how to get names of streets the user intersects?
> The maneuver.cross_street_names() seems to be here for it, but it doesn't seem to be set up anywhere.
> Also the tripleg_intersecting_edge class has no name attribute.

@grzezlo 
Currently, we only store a name consistency boolean flag to know if the intersecting edge has a consistent name with the path edge. The `maneuver.cross_street_names()` is only populated when we make a double-digitized u-turn

**Comment by grzezlo** [comment]
I made some progress, but also faced a strange error I can't solve.
After extending definition of IntersectingEdge in trip.proto:
```
repeated StreetName name = 9;           // street names
```
and copying appropriate class methods, I can set up the intersecting edge street names.
Further, passing the graphtile as additional parameter from TripLegBuilder::Build to AddIntersectingEdges, and finally to AddTripIntersectingEdge in triplegbuilder_utils.h, I can set the intersecting edge names using below block of code.
After successful compilation, and running some test routes, I get in log properly set names of intersecting edges, but finally I get error: [31;1m[ERROR][0m 400::GetNamesAndTypes: offset exceeds size of text list

No idea, what am I doing wrong and how to get the names in proper way?

The added block of code in AddTripIntersectingEdge (triplegbuilder_utils.h, line411) is as follows:
```
  if (controller.attributes.at(kEdgeNames)) {
    // Get the edgeinfo
    auto edgeinfo = graphtile->edgeinfo(intersecting_de);
    auto names_and_types = edgeinfo.GetNamesAndTypes();
    itersecting_edge->mutable_name()->Reserve(names_and_types.size());
    for (const auto& name_and_type : names_and_types) {
      auto* trip_edge_name = itersecting_edge->mutable_name()->Add();
      trip_edge_name->set_value(name_and_type.first);
      std::cout << "name: " << name_and_type.first << "\n";
      trip_edge_name->set_is_route_number(name_and_type.second);
    }
  }
```

**Comment by grzezlo** [comment]
After extending definition of IntersectingEdge in trip.proto, and cloning appropriate class methods, I can set up the intersecting edge street names in tripleg_builder and output in Narrative Builder.
Everything works fine, thanks for your advices.

**Comment by PhilipNgo** [comment]
@grzezlo I stumbled across this and thought it was very interesting, you don't happen to still have the code to produce the interesting result you mentioned?

**Comment by grzezlo** [comment]
Thanks @PhilipNgo for interest in this subject.
In fact, I have the modified Valhalla, which is successfully used by blind users of our company's app [Seeing Assistant Move](http://seeingassistant.tt.com.pl/).
It announces the streets which user should cross going through the route, not only those he or she should turn in.
The traffic lights on the crossings are also announced, and, last but not least, the tunnels, stairs, and similar, if possible.
Not sure if it would make sense to incorporate my modifications into official Valhalla, as it required to extend the data protobuf definitions, which could make the data storage less efficient without any added value for non-blind users.
Also not sure, if this issue should be closed, or not...

**Comment by nilsnolde** [comment]
fwiw, I think this is very interesting and would be a great addition. in case you're allowed and have the time to do so, I'd love to see a PR to review if it's feasible to include in master.

**Comment by PhilipNgo** [comment]
I agree with nilsnolde, it would be a great addition if possible. We are currently developing a similar app for guiding the blind, and we are exploring different approaches. This seems like an interesting path to dig deeper into.

**Comment by kevinkreiser** [comment]
@grzezlo Exactly as the others said if you can pr the work back to the repository we can help you with how to store the data or any other issues you may not have been sure about fitting into our library design

**Comment by grzezlo** [comment]
Sorry for the delay due to decision processes and vacation time.
Finally I have permission to publish our Valhalla modifications, so I'll prepare a pull-request with it.
It'll be great, that blind users mode will became a part of main code, so will be updated without merging it, and possibly improved by others interested.
Thanks @kevinkreiser for your help offer, I'll need it for sure.

**Comment by kevinkreiser** [comment]
@grzezlo that is fantastic news, i look forward to it!
